R/app-driver.R

```diff
@@ -989,28 +989,104 @@ AppDriver <- R6Class( # nolint
+    #' two images fail a [`testthat::compare_file_binary`] comparison.
+    #'
+    #' Which value should I use? Threshold values values below 5 help deter
+    #' false-positive screenshot comparisons (such as inconsistent rounded
+    #' corners). Larger values in the 10s and 100s will help find _real_
+    #' changes. However, not all values are one size fits all and you will need
+    #' to play with a threshold that fits your needs.
+    #' @param kernel_size Parameter supplied to [`compare_screenshot_threshold()`]
+    #' when using the default `compare` method. The `kernel_size` represents the
+    #' height and width of the convolution kernel applied to the pixel
+    #' differences. This integer-like value should be relatively small.
+    #' @param quiet Parameter supplied to [`compare_screenshot_threshold()`]
+    #' when using the default `compare` method. If `FALSE`, diagnostic
+    #' information will be presented when the computed value is larger than a
+    #' non-`NULL` `threshold` value.
     #' @examples
     #' \dontrun{
+    #' # These example lines should be performed in a `./tests/testthat`
+    #' # test file so that snapshot files can be properly saved
+    #'
     #' app_path <- system.file("examples/01_hello", package = "shiny")
     #' app <- AppDriver$new(app_path, variant = platform_variant())
     #'
-    #' # Expect a full size screenshot
+    #' # Expect a full size screenshot to be pixel perfect
     #' app$expect_screenshot()
     #'
-    #' # Very brittle test
+    #' # Images are brittle when containing plots
     #' app$expect_screenshot(selector = "#distPlot")
+    #'
+    #' # Test with more threshold in pixel value differences
+    #' # Helps with rounded corners
+    #' app$expect_screenshot(threshold = 10)
+    #'
+    #' # Equivalent expectations
+    #' app$expect_screenshot() # default
+    #' app$expect_screenshot(threshold = NULL)
+    #' app$expect_screenshot(compare = testthat::compare_file_binary)
+    #' expect_snapshot_file(
+    #'   app$get_screenshot(),
+    #'   variant = app$get_variant(),
+    #'   compare = testthat::compare_file_binary
+    #' )
+    #'
+    #' # Equivalent expectations
+    #' app$expect_screenshot(threshold = 3, kernel_size = 5)
+    #' app$expect_screenshot(compare = function(old, new) {
+    #'   compare_screenshot_threshold(
+    #'     old, new,
+    #'     threshold = 3,
+    #'     kernel_size = 5
+    #'   )
+    #' })
+    #' expect_screenshot_file(
+    #'   app$get_screenshot(),
+    #'   variant = app$get_variant(),
+    #'   compare = function(old, new) {
+    #'     compare_screenshot_threshold(
+    #'       old, new,
+    #'       threshold = 3,
+    #'       kernel_size = 5
+    #'     )
+    #'   }
+    #' )
     #' }
     expect_screenshot = function(
       ...,
+      threshold = NULL,
+      kernel_size = 5,
       screenshot_args = missing_arg(),
       delay = missing_arg(),
       selector = missing_arg(),
+      compare = missing_arg(),
```

Comment:
For future reference, here's another possibility for the API.

We could remove the `threshold` and `kernel_size` params from this function, and the default value for `compare` would be:

```R
  compare = testthat::compare_file_binary
```

And then if people want to use the fuzzy comparison, they would do something like:

```R
  compare = fuzzy_comparer(threshold = 3, kernel_size = 5)

  # Or maybe those values could use those as defaults
  compare = fuzzy_comparer()
```

The benefit is that it's a more general API, and provides a cleaner interface for plugging in other comparison functions. However, it seems unlikely anyone would ever use this capability, and it requires people to know about the `fuzzy_comparer` function.